### ISSUE TITLE
Code

### ISSUE THREAD

**Issue by ZfCai** [PXinTao/XYW-Net#1]
Can you share your code? I would like to use your method as the comparative experiment in my new paper
Thank you very much!!!

**Comment by PXinTao** [comment]
I'm glad to hear that you're interested in my method for your research. While I'm willing to share my code, it's important to note that it's not currently well-organized with comments and clear logic, and it would take some time to tidy up. Could you specify which part you're interested in? I can try to prepare that specific section for you.

**Comment by PXinTao** [comment]
> Can you share your code? I would like to use your method as the comparative experiment in my new paper Thank you very much!!!
I'm glad to hear that you're interested in my method for your research. While I'm willing to share my code, it's important to note that it's not currently well-organized with comments and clear logic, and it would take some time to tidy up. Could you specify which part you're interested in? I can try to prepare that specific section for you.

**Comment by ZfCai** [comment]
> > Can you share your code? I would like to use your method as the comparative experiment in my new paper Thank you very much!!!
> > I'm glad to hear that you're interested in my method for your research. While I'm willing to share my code, it's important to note that it's not currently well-organized with comments and clear logic, and it would take some time to tidy up. Could you specify which part you're interested in? I can try to prepare that specific section for you.

I'm from Hangzhou Dianzi University, our laboratory also studies edge detection tasks based on biological vision. I found your paper on Google and was very interested, but unfortunately, my coding ability is limited and cannot be reproduced. So it's presumptuous of me to leave this message for you to inquire the code of this paper.

**Comment by PXinTao** [comment]
> > > Can you share your code? I would like to use your method as the comparative experiment in my new paper Thank you very much!!!
> > > I'm glad to hear that you're interested in my method for your research. While I'm willing to share my code, it's important to note that it's not currently well-organized with comments and clear logic, and it would take some time to tidy up. Could you specify which part you're interested in? I can try to prepare that specific section for you.
> 
> I'm from Hangzhou Dianzi University, our laboratory also studies edge detection tasks based on biological vision. I found your paper on Google and was very interested, but unfortunately, my coding ability is limited and cannot be reproduced. So it's presumptuous of me to leave this message for you to inquire the code of this paper.

X Cell：
class Xc1x1(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(Xc1x1, self).__init__()

        self.Xcenter = nn.Conv2d(in_channels, out_channels, kernel_size=1)
        self.Xcenter_relu = nn.ReLU(inplace=True)
        self.Xsurround = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1, groups=in_channels)
        self.conv1_1 = nn.Conv2d(out_channels,out_channels,kernel_size=1)
        self.Xsurround_relu = nn.ReLU(inplace=True)


        self.in_channels = in_channels

    def forward(self, input):
        xcenter = self.Xcenter_relu(self.Xcenter(input))
        xsurround = self.Xsurround_relu(self.Xsurround(input))
        xsurround = self.conv1_1(xsurround)
        x = xsurround - xcenter
        return x


Y Cell：
class Yc1x1(nn.Module):
    def __init__(self, in_channels, out_channels):
        super(Yc1x1, self).__init__()
        self.Ycenter = nn.Conv2d(in_channels, out_channels, kernel_size=1)
        self.Ycenter_relu = nn.ReLU(inplace=True)
        self.Ysurround = nn.Conv2d(in_channels, out_channels, kernel_size=5, padding=4, dilation=2, groups=in_channels)
        self.conv1_1 = nn.Conv2d(out_channels, out_channels, kernel_size=1)
        self.Ysurround_relu = nn.ReLU(inplace=True)



    def forward(self, input):
        ycenter = self.Ycenter_relu(self.Ycenter(input))
        ysurround = self.Ysurround_relu(self.Ysurround(input))
        ysurround = self.conv1_1(ysurround)
        y = ysurround - ycenter

        return y
       
      W Cell ：
      class W(nn.Module):
    def __init__(self, inchannel, outchannel,stride=1):
        super(W, self).__init__()

        self.h = nn.Conv2d(inchannel, inchannel, kernel_size=(1, 3), padding=(0, 1),groups=inchannel)
        self.v = nn.Conv2d(inchannel, inchannel, kernel_size=(3, 1), padding=(1, 0),groups=inchannel)

        self.convh_1 = nn.Conv2d(in_channels= inchannel, out_channels=inchannel, kernel_size=1, padding=0,bias=False)
        self.convv_1 = nn.Conv2d(in_channels= inchannel, out_channels=outchannel, kernel_size=1, padding=0,bias=False)
        self.relu = nn.ReLU()

    def forward(self, x):

        h = self.relu(self.h(x))
        h = self.convh_1(h)

        v = self.relu(self.v(h))
        v = self.convv_1(v)

class XYW_S(nn.Module):
    def __init__(self, inchannel, outchannel, stride=1):
        super(XYW_S, self).__init__()
        self.stride = stride
        self.y_c = Yc1x1(inchannel , outchannel)
        self.x_c = Xc1x1(inchannel , outchannel)
        self.w = W(inchannel, outchannel)

        self.relu = nn.ReLU()
       
    def forward(self, x):

        xc = self.x_c(x)
        yc = self.y_c(x)
        w = self.w(x)
        return xc, yc, w
        
class XYW(nn.Module):
    def __init__(self, inchannel, outchannel, stride=1):
        super(XYW, self).__init__()
        self.y_c = Yc1x1(inchannel, outchannel)
        self.x_c = Xc1x1(inchannel, outchannel)
        self.w = W(inchannel, outchannel)

        self.relu = nn.ReLU()



    def forward(self, xc, yc, w):

        xc = self.x_c(xc)
        yc = self.y_c(yc)
        w = self.w(w)

        return xc, yc, w


class XYW_E(nn.Module):       
    def __init__(self, inchannel, outchannel):
        super(XYW_E, self).__init__()
        self.y_c = Yc1x1(inchannel , outchannel)
        self.x_c = Xc1x1(inchannel , outchannel)
        self.w = W(inchannel, outchannel)

        self.relu = nn.ReLU()
    
    def forward(self, xc, yc, w):
        xc = self.x_c(xc)
        yc = self.y_c(yc)
        w = self.w(w)

        return xc+yc+w
        
       The above is my bio-inspired module, combined with the description in the paper, I hope this can be of help to you. After this period, I will also rush to update the code, wishing you all the best.

**Comment by ZfCai** [comment]
> > > > Can you share your code? I would like to use your method as the comparative experiment in my new paper Thank you very much!!!
> > > > I'm glad to hear that you're interested in my method for your research. While I'm willing to share my code, it's important to note that it's not currently well-organized with comments and clear logic, and it would take some time to tidy up. Could you specify which part you're interested in? I can try to prepare that specific section for you.
> > 
> > 
> > I'm from Hangzhou Dianzi University, our laboratory also studies edge detection tasks based on biological vision. I found your paper on Google and was very interested, but unfortunately, my coding ability is limited and cannot be reproduced. So it's presumptuous of me to leave this message for you to inquire the code of this paper.
> 
> X Cell： class Xc1x1(nn.Module): def **init**(self, in_channels, out_channels): super(Xc1x1, self).**init**()
> 
> ```
>     self.Xcenter = nn.Conv2d(in_channels, out_channels, kernel_size=1)
>     self.Xcenter_relu = nn.ReLU(inplace=True)
>     self.Xsurround = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1, groups=in_channels)
>     self.conv1_1 = nn.Conv2d(out_channels,out_channels,kernel_size=1)
>     self.Xsurround_relu = nn.ReLU(inplace=True)
> 
> 
>     self.in_channels = in_channels
> 
> def forward(self, input):
>     xcenter = self.Xcenter_relu(self.Xcenter(input))
>     xsurround = self.Xsurround_relu(self.Xsurround(input))
>     xsurround = self.conv1_1(xsurround)
>     x = xsurround - xcenter
>     return x
> ```
> 
> Y Cell： class Yc1x1(nn.Module): def **init**(self, in_channels, out_channels): super(Yc1x1, self).**init**() self.Ycenter = nn.Conv2d(in_channels, out_channels, kernel_size=1) self.Ycenter_relu = nn.ReLU(inplace=True) self.Ysurround = nn.Conv2d(in_channels, out_channels, kernel_size=5, padding=4, dilation=2, groups=in_channels) self.conv1_1 = nn.Conv2d(out_channels, out_channels, kernel_size=1) self.Ysurround_relu = nn.ReLU(inplace=True)
> 
> ```
> def forward(self, input):
>     ycenter = self.Ycenter_relu(self.Ycenter(input))
>     ysurround = self.Ysurround_relu(self.Ysurround(input))
>     ysurround = self.conv1_1(ysurround)
>     y = ysurround - ycenter
> 
>     return y
>    
>   W Cell ：
>   class W(nn.Module):
> def __init__(self, inchannel, outchannel,stride=1):
>     super(W, self).__init__()
> 
>     self.h = nn.Conv2d(inchannel, inchannel, kernel_size=(1, 3), padding=(0, 1),groups=inchannel)
>     self.v = nn.Conv2d(inchannel, inchannel, kernel_size=(3, 1), padding=(1, 0),groups=inchannel)
> 
>     self.convh_1 = nn.Conv2d(in_channels= inchannel, out_channels=inchannel, kernel_size=1, padding=0,bias=False)
>     self.convv_1 = nn.Conv2d(in_channels= inchannel, out_channels=outchannel, kernel_size=1, padding=0,bias=False)
>     self.relu = nn.ReLU()
> 
> def forward(self, x):
> 
>     h = self.relu(self.h(x))
>     h = self.convh_1(h)
> 
>     v = self.relu(self.v(h))
>     v = self.convv_1(v)
> ```
> 
> class XYW_S(nn.Module): def **init**(self, inchannel, outchannel, stride=1): super(XYW_S, self).**init**() self.stride = stride self.y_c = Yc1x1(inchannel , outchannel) self.x_c = Xc1x1(inchannel , outchannel) self.w = W(inchannel, outchannel)
> 
> ```
>     self.relu = nn.ReLU()
>    
> def forward(self, x):
> 
>     xc = self.x_c(x)
>     yc = self.y_c(x)
>     w = self.w(x)
>     return xc, yc, w
> ```
> 
> class XYW(nn.Module): def **init**(self, inchannel, outchannel, stride=1): super(XYW, self).**init**() self.y_c = Yc1x1(inchannel, outchannel) self.x_c = Xc1x1(inchannel, outchannel) self.w = W(inchannel, outchannel)
> 
> ```
>     self.relu = nn.ReLU()
> 
> 
> 
> def forward(self, xc, yc, w):
> 
>     xc = self.x_c(xc)
>     yc = self.y_c(yc)
>     w = self.w(w)
> 
>     return xc, yc, w
> ```
> 
> class XYW_E(nn.Module): def **init**(self, inchannel, outchannel): super(XYW_E, self).**init**() self.y_c = Yc1x1(inchannel , outchannel) self.x_c = Xc1x1(inchannel , outchannel) self.w = W(inchannel, outchannel)
> 
> ```
>     self.relu = nn.ReLU()
> 
> def forward(self, xc, yc, w):
>     xc = self.x_c(xc)
>     yc = self.y_c(yc)
>     w = self.w(w)
> 
>     return xc+yc+w
>     
>    The above is my bio-inspired module, combined with the description in the paper, I hope this can be of help to you. After this period, I will also rush to update the code, wishing you all the best.
> ```

Thanks sincerely

**Comment by ZfCai** [comment]
我根据您给出的代码尝试复现，但是在经过20epoch的训练之后，所得到的测试图片有异常多的纹理，且轮廓很淡，请问您知道可能是什么原因导致的嘛？
![image](https://github.com/PXinTao/XYW-Net/assets/147697478/e52fdffc-b9ed-42ba-a961-a14d4d26c982)
